pkg/build/gather-build-info.go

```diff
@@ -0,0 +1,152 @@
+package build
+
+import (
+	"fmt"
+	"github.com/bugsnag/bugsnag-cli/pkg/android"
+	"github.com/bugsnag/bugsnag-cli/pkg/options"
+	"github.com/bugsnag/bugsnag-cli/pkg/utils"
+)
+
+type SourceControl struct {
+	Provider   string `json:"provider,omitempty"`
+	Repository string `json:"repository,omitempty"`
+	Revision   string `json:"revision,omitempty"`
+}
+
+type CreateBuildInfo struct {
+	ApiKey            string            `json:"apiKey,omitempty"`
+	AppVersionCode    string            `json:"appVersionCode,omitempty"`
+	AppBundleVersion  string            `json:"appBundleVersion,omitempty"`
+	SourceControl     SourceControl     `json:"sourceControl,omitempty"`
+	BuilderName       string            `json:"builderName,omitempty"`
+	ReleaseStage      string            `json:"releaseStage,omitempty"`
+	AppVersion        string            `json:"appVersion,omitempty"`
+	AutoAssignRelease *bool             `json:"autoAssignRelease,omitempty"`
+	MetaData          map[string]string `json:"metadata,omitempty"`
+}
+
+func (opts CreateBuildInfo) Override(base CreateBuildInfo) CreateBuildInfo {
+	return CreateBuildInfo{
+		ApiKey:           utils.ThisOrThat(opts.ApiKey, base.ApiKey).(string),
+		AppVersionCode:   utils.ThisOrThat(opts.AppVersionCode, base.AppVersionCode).(string),
+		AppBundleVersion: utils.ThisOrThat(opts.AppBundleVersion, base.AppBundleVersion).(string),
+		SourceControl: SourceControl{
+			Provider:   utils.ThisOrThat(opts.SourceControl.Provider, base.SourceControl.Provider).(string),
+			Repository: utils.ThisOrThat(opts.SourceControl.Repository, base.SourceControl.Repository).(string),
+			Revision:   utils.ThisOrThat(opts.SourceControl.Revision, base.SourceControl.Revision).(string),
+		},
+		BuilderName:       utils.ThisOrThat(opts.BuilderName, base.BuilderName).(string),
+		ReleaseStage:      utils.ThisOrThat(opts.ReleaseStage, base.ReleaseStage).(string),
+		AppVersion:        utils.ThisOrThat(opts.AppVersion, base.AppVersion).(string),
+		AutoAssignRelease: utils.ThisOrThatBool(opts.AutoAssignRelease, base.AutoAssignRelease),
+		MetaData:          utils.ThisOrThat(opts.MetaData, base.MetaData).(map[string]string),
+	}
+}
+
+func (opts CreateBuildInfo) Validate() error {
+	if opts.ApiKey == "" {
+		return fmt.Errorf("no API key provided")
+	}
+
+	if opts.AppVersion == "" {
+		return fmt.Errorf("Missing app version, please provide this via the command line options")
+	}
+
+	if opts.SourceControl.Repository == "" || opts.SourceControl.Revision == "" {
+		return fmt.Errorf("Missing Source Control Repository or Revision")
+	}
+
+	return nil
```

Comment:
We should also validate that provider (if given) is one of the permitted values:

* github
* github-enterprise
* bitbucket
* bitbucket-server
* gitlab
* gitlab-onpremise